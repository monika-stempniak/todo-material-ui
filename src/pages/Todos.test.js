import React from 'react';
import axios from 'axios';
import { render, fireEvent } from '@testing-library/react';
import { waitForDomChange } from '@testing-library/dom';

import { todos } from '../lib/todos-mock';
import { Todos } from '.';

jest.mock('axios');

describe('Todos component', () => {
  beforeEach(() => {
    axios.get.mockResolvedValueOnce({data: todos});
  });
  afterEach(() => {
    axios.get.mockReset();
  });

  it("matches snapshot", async () => {
    const { asFragment } = render(<Todos />);

    await waitForDomChange();
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render list of todos', async () => {
    const { getByTestId } = render(<Todos />);
    expect(getByTestId('todos-loading').innerHTML).toMatch('Loading...');

    await waitForDomChange();

    expect(axios.get).toHaveBeenCalledTimes(1);
    const todoList = getByTestId('todos-list');
    expect(todoList).toBeTruthy();
    expect(todoList.children.length).toBe(3);
  })

  test('should allow adding new todos', async () => {
    const { getByTestId, debug, getByLabelText, getByText, getAllByTestId } = render(<Todos />);

    await waitForDomChange();

    const todoInput = getByLabelText('Add todo');
    const todo = 'Learn React';
    fireEvent.change(todoInput, { target: { value: todo } });
    expect(todoInput.value).toBe(todo);

    fireEvent.submit(getByTestId('add-button'));
    // debug();
    expect(todoInput.value).toBe('');
    const todoList = getByTestId('todos-list');
    expect(todoList.children.length).toBe(4);
    expect(getByText(todo)).toBeTruthy();
    expect(getAllByTestId('todo')[0].innerHTML).toContain(todo);
  });
});

describe('Todos component - error test', () => {
  test('should display an error if request fails', async () => {
    const error = 'Async error';
    axios.get.mockRejectedValue(new Error(error));

    const { container, queryByTestId } = render(<Todos />);

    await waitForDomChange();

    expect(axios.get).toHaveBeenCalledTimes(1);
    expect(container).toHaveTextContent(`Something went wrong! ${error}`);

    const todoList = document.querySelector('[data-testid="todos-list"]');
    expect(todoList).not.toBeInTheDocument();
  });
});