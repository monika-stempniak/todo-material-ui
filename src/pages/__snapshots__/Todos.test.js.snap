// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root makeStyles-root-3 MuiContainer-fixed MuiContainer-maxWidthLg"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
          data-testid="title"
        >
          Todos
        </h5>
        <form
          class="makeStyles-form-144"
        >
          <div
            class="MuiFormControl-root MuiTextField-root makeStyles-textField-145 MuiFormControl-marginNormal"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
              data-shrink="false"
              for="standard-name"
            >
              Add todo
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                id="standard-name"
                type="text"
                value=""
              />
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root makeStyles-button-146 MuiButton-contained"
            data-testid="add-button"
            tabindex="0"
            type="submit"
          >
            <span
              class="MuiButton-label"
            >
              Add
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </form>
      </div>
      <ul
        class="MuiList-root makeStyles-list-2 MuiList-padding"
        data-testid="todos-list"
      >
        <li
          class="MuiListItem-root MuiListItem-gutters"
        >
          <p
            class="MuiTypography-root makeStyles-todo-235 makeStyles-todo-238 MuiTypography-body1"
            data-testid="todo"
          >
            delectus aut autem
          </p>
          <button
            class="MuiButtonBase-root MuiIconButton-root makeStyles-checkIcon-236"
            data-testid="complete"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 7.58L10 14.17l-3.59-3.58L5 12l5 5 8-8zM12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </li>
        <li
          class="MuiListItem-root MuiListItem-gutters"
        >
          <p
            class="MuiTypography-root makeStyles-todo-235 makeStyles-todo-268 MuiTypography-body1"
            data-testid="todo"
          >
            quis ut nam facilis et officia qui
          </p>
          <button
            class="MuiButtonBase-root MuiIconButton-root makeStyles-checkIcon-236"
            data-testid="complete"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 7.58L10 14.17l-3.59-3.58L5 12l5 5 8-8zM12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </li>
        <li
          class="MuiListItem-root MuiListItem-gutters"
        >
          <p
            class="MuiTypography-root makeStyles-todo-235 makeStyles-todo-269 MuiTypography-body1"
            data-testid="todo"
          >
            fugiat veniam minus
          </p>
          <button
            class="MuiButtonBase-root MuiIconButton-root makeStyles-checkIcon-236"
            data-testid="complete"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 7.58L10 14.17l-3.59-3.58L5 12l5 5 8-8zM12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
